brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
 import com.stripe.brushfire._
 import com.twitter.algebird._
 
-case class Trainer[K: Ordering, V, T: Monoid](
-    trainingData: Iterable[Instance[K, V, T]],
-    sampler: Sampler[K],
-    trees: List[Tree[K, V, T]]) {
 
-  private def updateTrees(fn: (Tree[K, V, T], Int, Map[LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]]) => Tree[K, V, T]): Trainer[K, V, T] = {
-    val newTrees = trees.zipWithIndex.par.map {
-      case (tree, index) =>
-        val byLeaf =
-          trainingData.flatMap { instance =>
-            val repeats = sampler.timesInTrainingSet(instance.id, instance.timestamp, index)
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                1.to(repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
-    }
-    copy(trees = newTrees.toList)
+trait TrainingStep[K,V,T] {
+
+  type K1
```

Comment:
(Just sort of trying to see if I understand these types)

So, this is some key that can be used to segment the initial per-leaf values? I think this could technically be replaced by using a map in `V1`, so is the main reason for this to optimize our m/r steps (since it is reasonable to expect `|K1| = O(|features|)`)?